### PR TITLE
Shutdown provider

### PR DESCRIPTION
**Pull request by maaktweluit**
Add the option to shutdown the provider service at any time

- Made `ProviderAppClient` use `self._service` like requestor
- Added a Wrapper as kill switch in the `ProviderAppClient`
- Added try blocks on the server to fire on the shutdown event
- Added `TaskApiService.running()` to check if the service has already shut down

### CODE REVIEW

File: python/golem_task_api/client.py [hunk 1]
```diff
@@ -134,43 +144,55 @@ def __init__(
...
-        )
-        return ProviderAppStub(
+            port: int = DEFAULT_PORT,
+    ) -> None:
+        self._service = service
+        host, port = service.start(f'provider {port}', port)
+        self._golem_app = ProviderAppStub(
             Channel(host, port, loop=asyncio.get_event_loop()),
         )
+        self._kill_switch = Wrapper()
```
Comment (Wiezzel) [inline python/golem_task_api/client.py]: Don't we need this kind of switch for `RequestorAppClient` as well?

File: python/golem_task_api/server.py [hunk 2]
```diff
@@ -156,7 +156,12 @@ def __init__(
     @forward_exceptions()
     async def RunBenchmark(self, stream):
         await stream.recv_message()
-        score = await self._handler.run_benchmark(self._work_dir)
+        try:
+            score = await self._handler.run_benchmark(self._work_dir)
+        except Exception as e:
```
Comment (Wiezzel) [inline python/golem_task_api/server.py]: Isn't `@forward_exception` already doing this job?

File: python/golem_task_api/client.py (hunk 1 above)
Comment (maaktweluit) [reply python/golem_task_api/client.py]: yes, will make an issue if i cant fix that this sprint :)

File: python/golem_task_api/server.py (hunk 2 above)
Comment (maaktweluit) [reply python/golem_task_api/server.py]: i had some issues with showing the errors so i added this block

we can still use it for setting status or executing finally ( see TODO next line ) or remove it and let `@forward_exception` handle it

File: python/golem_task_api/client.py [hunk 3]
```diff
@@ -134,43 +144,55 @@ def __init__(
...
+        except Exception as e:
+            print('TODO: Set status / use finally or remove this try block')
+            print('ERROR in RunBenchmark:', e)
+            raise
+        await self._service.wait_until_shutdown_complete()
         return reply.score
+
+    async def shutdown(self) -> None:
+        if not self._kill_switch.cancelled:
+            self._kill_switch.cancel(Exception("shutdown"))
```
Comment (Wiezzel) [inline python/golem_task_api/client.py]: Some more specific exception type (possibly a custom one) would be better here. Caller of `run_bechmark` or `compute` might want to handle cancellations differently than actual errors.

File: python/golem_task_api/server.py (hunk 2 above)
Comment (Wiezzel) [reply python/golem_task_api/server.py]: So, what do we do with this?

File: python/golem_task_api/client.py (hunk 3 above)
Comment (maaktweluit) [reply python/golem_task_api/client.py]: good one, will make it a custom exception

File: python/golem_task_api/server.py (hunk 2 above)
Comment (maaktweluit) [reply python/golem_task_api/server.py]: for now i'll remove it, we can always add it back when we want to set a status

Comment (Wiezzel) [reply python/golem_task_api/server.py]: OK, fine.

File: python/golem_task_api/client.py (hunk 1 above)
Comment (maaktweluit) [reply python/golem_task_api/client.py]: CLAY-410

**Pull request merged**